lib/galaxy/authnz/custos_authnz.py

```diff
@@ -230,3 +247,12 @@ def _get_verify_param(self):
             return self.config['ca_bundle']
         else:
             return self.config['verify_ssl']
+
+    def generate_username(self, trans, email):
+        temp_username = email.split('@')[0] #username created from username portion of email
+        count = 0
+
+        while (trans.sa_session.query(trans.app.model.User).filter_by(username=(temp_username+str(count))).first()):
+        #if username already exists in database, append integer and iterate until unique username found
```

Comment:
The number of near-identical usernames could be high in some cases ("example#", "test#", "john.smith#"...). Issuing sequential queries to the db to retrieve each one would be unnecessarily expensive. I'm not familiar with the context, so it could be a non-issue. Otherwise, you could maybe retrieve all partial matches in one query and generate the next username+count based on that result?